### ISSUE TITLE
Custom model Inference Instance segmentation 

### ISSUE THREAD

**Issue by bilaloumehdi** [ultralytics/yolov5#11506]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have trained the yolov5 on my custom dataset for instance segmentation with python. then I used the model for inference and I am getting a list of Tensors of size  :  "torch.Size([1, 10647, 40])".

I am wondering how to extract the information from these Tensors (conf, xyxy, mask,...) . 

Thank you.

**Comment by glenn-jocher** [comment]
Hello @Bilaloum, 

To extract the information from the output tensor, you can use the `pred` attribute returned by the `detect` method. `pred` is a list of length equal to the batch size and each element represents an image in the batch. 

Each image element in `pred` is a tensor of shape `(n, 6)` representing `n` number of detected instances. The columns correspond to `(x1, y1, x2, y2, confidence, class)`. 

To perform instance segmentation on the detected objects, you can retrieve the `mask` tensor which is of shape `(n, 1, H, W)` and represents a binary mask for each object. Note that instance segmentation is only possible if your model was trained using mask labels.

Hope this helps!

**Comment by bilaloumehdi** [comment]
Hello @glenn-jocher,

- So I need to implement the detect method that returns a  `pred ` attribute because my model does not have the `detect `method.
here is it the code I am using: 
 ` model = torch.hub.load('ultralytics/yolov5', 'custom', path='best.pt', force_reload=True)`
 `results = model(img_tensor)`

**the result is a list of Tensors of size ([1, 10647, 40]) , I don't know what do next to get the results as list of objects containing the previous mentioned information**

**I have trained my model using `segment/train.py`  on a custom carob datasets labeled and  annotated using Rboflow SAM functionality**.

• does this Warning means that I need to do something about:  "YOLOv5 SegmentationModel is not yet AutoShape compatible. You will not be able to run inference with this model."

Thank you for your help.

**Comment by glenn-jocher** [comment]
Hello @Bilaloum,

You don't need to implement the `detect` method yourself. Instead, you should use the `model` object's `forward() `method to get the output of the model. For instance segmentation, this will return a list of tensors of size `(N, C, H, W)`, where `N` is the number of instances detected per image, `C` is the number of classes and `H`, `W` are the height and width of the mask.

To extract information regarding the detected instances like bounding box information (x1, y1, x2, y2), class confidence scores, class IDs, and mask tensors, you can use YOLOv5's built-in `postprocess()` method. For example:

```
from yolov5.utils.parsing import get_classes
from yolov5.utils.torch_utils import select_device
from yolov5.models.experimental import attempt_load
from yolov5.utils.datasets import LoadImages
from yolov5.models.yolo import Detect

weights = 'best.pt'
img_size = 512
conf_thres = 0.3
iou_thres = 0.5
device = select_device('cpu')
model = attempt_load(weights, map_location=device)

img = 'example.jpg'
img, _ = LoadImages(img_size=img_size, img_formats=['.jpg', '.jpeg', '.png'])(img)
pred = model(img)[0]

classes = get_classes('custom.names')
coco89 = False
agnostic_nms = False
detections = Detect(conf_thres=conf_thres, iou_thres=iou_thres, agnostic_nms=agnostic_nms,
                    augment=False, classes=coco89, pad=0.5)

# Convert predictions to (x, y, w, h), class, confidence, segmentation format
results = detections(img, pred)
# Render segmentation masks
results.render()
# Print predictions
print(results.pandas().xyxy[0])
```

Regarding the warning message you mentioned, it means that the model is not compatible with YOLOv5's new AutoShape API. It's just a warning and should not affect the model's performance. However, we recommend updating your model to the new API to take advantage of its simplicity and compatibility with different input sizes.

I hope this helps! Let us know if you have any further questions.

**Comment by bilaloumehdi** [comment]
Hello @glenn-jocher , 

I have tried this approach and it's not working. because it not finding `yolov5.utils.parsing ` and   `yolov5.utils.datasets`.
Could you give me an example using

- `torch.hub.load() ` to load my custom model 

- process the image input to add the batch size and then pass it to the model 

- extract from list of tensors the predection information such as the mask to calculate the mask area at the end.

Thank for your help sir.

**Comment by glenn-jocher** [comment]
Hello @Bilaloum,

I apologize for the confusion in my previous response. Since you are using `torch.hub.load()` to load your custom model, you can use the following steps to obtain instance masks from the output of your model:

```
import torch
import numpy as np

# Load the model using torch.hub
model = torch.hub.load('ultralytics/yolov5', 'custom', path='path/to/custom/model.pt', source='local')

# Set the device to use (GPU or CPU)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# Transform the input image tensor to a batched tensor shape (N, C, H, W)
img_tensor_batched = img_tensor.unsqueeze(0)

# Move the input tensor to the device
img_tensor_batched = img_tensor_batched.to(device)

# Pass the input tensor to the model to obtain the list of detected instances masks
res = model(img_tensor_batched, size=512)
instances = res.pandas().xyxy[0]

# Extract instance class and instance mask tensors from the list of detected instances
masks = []
class_ids = []
for instance in instances:
    # Extract the cropped image patch from the input image using the bbox coordinates
    x1, y1, x2, y2 = instance[:4].detach().cpu().numpy().astype(np.int)
    cropped_img = img_tensor[:, y1:y2, x1:x2].unsqueeze(0).to(device)
    
    # Run the cropped image patch through the model to obtain the mask tensor
    mask = model(cropped_img, size=512).pred[0].squeeze()
    masks.append(mask)
    class_ids.append(instance[-1].detach().cpu().numpy().astype(np.int))
    
# Calculate the area of each instance mask
areas = [torch.sum(mask) for mask in masks]
```

In summary, this code loads the custom model using `torch.hub.load()`, transforms the input image tensor to a batched tensor shape, and passes it to the model to obtain the list of detected instance boxes. It then extracts the instance mask and class ID for each detected instance, and calculates the area of each instance mask. Finally, the areas are accumulated into a list named `areas`.

Let me know if this solves the problem.

**Comment by anewworl** [comment]
> Hello @Bilaloum,
> 
> You don't need to implement the `detect` method yourself. Instead, you should use the `model` object's `forward() `method to get the output of the model. For instance segmentation, this will return a list of tensors of size `(N, C, H, W)`, where `N` is the number of instances detected per image, `C` is the number of classes and `H`, `W` are the height and width of the mask.
> 
> To extract information regarding the detected instances like bounding box information (x1, y1, x2, y2), class confidence scores, class IDs, and mask tensors, you can use YOLOv5's built-in `postprocess()` method. For example:
> 
> ```
> from yolov5.utils.parsing import get_classes
> from yolov5.utils.torch_utils import select_device
> from yolov5.models.experimental import attempt_load
> from yolov5.utils.datasets import LoadImages
> from yolov5.models.yolo import Detect
> 
> weights = 'best.pt'
> img_size = 512
> conf_thres = 0.3
> iou_thres = 0.5
> device = select_device('cpu')
> model = attempt_load(weights, map_location=device)
> 
> img = 'example.jpg'
> img, _ = LoadImages(img_size=img_size, img_formats=['.jpg', '.jpeg', '.png'])(img)
> pred = model(img)[0]
> 
> classes = get_classes('custom.names')
> coco89 = False
> agnostic_nms = False
> detections = Detect(conf_thres=conf_thres, iou_thres=iou_thres, agnostic_nms=agnostic_nms,
>                     augment=False, classes=coco89, pad=0.5)
> 
> # Convert predictions to (x, y, w, h), class, confidence, segmentation format
> results = detections(img, pred)
> # Render segmentation masks
> results.render()
> # Print predictions
> print(results.pandas().xyxy[0])
> ```
> 
> Regarding the warning message you mentioned, it means that the model is not compatible with YOLOv5's new AutoShape API. It's just a warning and should not affect the model's performance. However, we recommend updating your model to the new API to take advantage of its simplicity and compatibility with different input sizes.
> 
> I hope this helps! Let us know if you have any further questions.

hello @glenn-jocher 

I see this code above and the line "results.render()" is quite unclear for me can you explain more about what it do and what is the data we get from this line.

**Comment by glenn-jocher** [comment]
Hello @anewworl,

The line `results.render()` is used to render the segmentation masks on the input image. It visualizes the segmentation masks overlaid on the original image, allowing you to see the masks of the detected instances. The rendered image could be useful for visual inspection or analysis.

As for the data obtained from this line, `results.pandas().xyxy[0]` gives a pandas DataFrame containing the bounding box coordinates `(x1, y1, x2, y2)`, the class label, confidence scores, and other information for each detected instance. The index `0` is used to access the first image's results in case there are multiple images processed.

Please let me know if you have any further questions or need additional clarification.

**Comment by anewworl** [comment]
Hello @glenn-jocher 

`[array([[[255,  56,  56],
        [255,  56,  56],
        [255,  56,  56],
        ...,
        [247, 247, 224],
        [246, 248, 225],
        [245, 247, 224]],

       [[255,  56,  56],
        [255,  56,  56],
        [255,  56,  56],
        ...,
        [245, 241, 224],
        [242, 238, 220],
        [249, 245, 227]],

       [[255,  56,  56],
        [255,  56,  56],
        [255,  56,  56],
        ...,
        [241, 235, 219],
        [243, 239, 221],
        [248, 244, 226]],

       ...,

       [[255,  56,  56],
        [255,  56,  56],
        [ 81,  35,  74],
        ...,
        [157, 179, 184],
        [154, 176, 181],
        [151, 174, 177]],

       [[255,  56,  56],
        [255,  56,  56],
        [119,  38,  70],
        ...,
        [158, 182, 186],
        [156, 178, 183],
        [154, 176, 181]],

       [[255,  56,  56],
        [255,  56,  56],
        [255,  56,  56],
        ...,
        [153, 183, 186],
        [152, 180, 184],
        [150, 178, 182]]], dtype=uint8)]`
So this is the result i get from the result.render() function, it likely seem that what i recieve is likely a RGB value. But i want to extract a list of `[x,y]` point so that i can use `cv2.polyline` in order to create a surrounding line of detected object. 
My question is that is there any way we can do it in yolov5.

**Comment by glenn-jocher** [comment]
@anewworl hello,

The result you obtained from `results.render()` represents an image with RGB values. If you want to extract a list of `[x, y]` points for each detected object to create a surrounding line using `cv2.polyline`, you can achieve it using the YOLOv5 library. Here's an example of how you can do it:

```python
import cv2
from yolov5.utils.parsing import get_classes
from yolov5.utils.torch_utils import select_device
from yolov5.models.experimental import attempt_load
from yolov5.utils.datasets import LoadImages
from yolov5.models.yolo import Detect

weights = 'best.pt'
img_size = 512
conf_thres = 0.3
iou_thres = 0.5
device = select_device('cpu')
model = attempt_load(weights, map_location=device)

img = 'example.jpg'
img, _ = LoadImages(img_size=img_size, img_formats=['.jpg', '.jpeg', '.png'])(img)
pred = model(img)[0]

classes = get_classes('custom.names')
coco89 = False
agnostic_nms = False
detections = Detect(conf_thres=conf_thres, iou_thres=iou_thres, agnostic_nms=agnostic_nms,
                    augment=False, classes=coco89, pad=0.5)

# Convert predictions to (x, y, w, h), class, confidence, segmentation format
results = detections(img, pred)
results.render()
boxes = results.pred[0][:, :4].detach().cpu().numpy()

# Create a surrounding line for each detected object
for box in boxes:
    x1, y1, x2, y2 = box
    pts = np.array([[x1, y1], [x1, y2], [x2, y2], [x2, y1]], np.int32)
    pts = pts.reshape((-1, 1, 2))
    cv2.polylines(img, [pts], True, (255, 0, 0), thickness=2)

cv2.imshow('Surrounding Lines', img)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

In this code, we load the YOLOv5 model, process the input image, and make predictions using `results = detections(img, pred)`. Then we access the bounding boxes using

**Comment by anewworl** [comment]
@glenn-jocher, thanks for your answer but maybe i am saying it not clear. 

![z4655634008936_cebdffa89b6da87ec50098618537ec39](https://github.com/ultralytics/yolov5/assets/104755734/308ec418-314c-443a-8915-b1c6a5b0a226)
![z4655672072479_061dcd4cbc2b24729999120219836617](https://github.com/ultralytics/yolov5/assets/104755734/0cccba23-d232-4c1d-9f56-aca3c7e937a8)


This is the result i am looking for so can we somehow extract the bounding points which is the list of `[x, y]` from `yolov5`. The method above he is using is the `contours` in `OpenCV` (1st pic), and the `segmentation` in `yolov8` so i wonder if the `yolov5` provided the same method as the `contours` or the `segmentation points` like in `yolov8` so that i can make a bounding line as closed as the object as possible with `cv2.polylines`. Because i know that `segmentation` model in `yolov5` is not autoshape yet.

**Comment by glenn-jocher** [comment]
@anewworl based on your query, it seems you're looking to extract the bounding points `[x, y]` from YOLOv5 for creating a bounding line similar to the contour or segmentation points in YOLOv8. YOLOv5 doesn't have a built-in method to directly extract contour or segmentation points like YOLOv8. However, you can still achieve your goal by using the bounding box coordinates provided by YOLOv5 and creating a surrounding line using `cv2.polylines`.

Here's an example of how you can create a bounding line similar to the contour using YOLOv5 and `cv2.polylines`:

```python
import cv2
from yolov5.utils.parsing import get_classes
from yolov5.utils.torch_utils import select_device
from yolov5.models.experimental import attempt_load
from yolov5.utils.datasets import LoadImages
from yolov5.models.yolo import Detect

weights = 'best.pt'
img_size = 512
conf_thres = 0.3
iou_thres = 0.5
device = select_device('cpu')
model = attempt_load(weights, map_location=device)

img = 'example.jpg'
img, _ = LoadImages(img_size=img_size, img_formats=['.jpg', '.jpeg', '.png'])(img)
pred = model(img)[0]

classes = get_classes('custom.names')
coco89 = False
agnostic_nms = False
detections = Detect(conf_thres=conf_thres, iou_thres=iou_thres, agnostic_nms=agnostic_nms,
                    augment=False, classes=coco89, pad=0.5)

# Convert predictions to (x, y, w, h), class, confidence, segmentation format
results = detections(img, pred)
results.render()
boxes = results.pred[0][:, :4].detach().cpu().numpy()

# Create a surrounding line for each detected object
for box in boxes:
    x1, y1, x2, y2 = box
    pts = np.array([[x1, y1], [x1, y2], [x2, y2], [x2, y1]], np.int32)
    pts = pts.reshape((-1, 1, 2))
    cv2.polylines(img, [pts], True, (255, 0, 0), thickness=

**Comment by ziyuuuuuu** [comment]
I have a question. I have used my custom-trained Yolov5 instance segmentation model and finished the inference process. but since I have two different classes that showed up at the same image, I am wondering if I can do a filter process to remain just the aimed class. I have used the supervision method (https://roboflow.com/how-to-filter-detections/yolov5) but it is not working well...

**Comment by anewworl** [comment]
@ziyuuuuuu Hm can you give more context about the case because the training step hold an important role in your case so we can discuss more . The case in the link you are given only work when the yolo can detect both at the time then you can make and if-else. But somehow if the yolo can only detect 1 out 2 object the method above is seem to not suitable anymore.

**Comment by glenn-jocher** [comment]
@ziyuuuuuu Sure, it would be helpful to have more information about your case to provide you with a more specific solution. Could you please provide more details about the issue you're facing with your custom-trained YOLOv5 instance segmentation model?

If the approach mentioned in the link you shared is not working for you, it could be because the method assumes that both classes are detected simultaneously. In cases where only one class is detected, this method may not be suitable.

Please provide further details about your training process, such as the dataset used, number of images per class, model configuration, and any other relevant information. This will allow us to better understand the issue and provide you with appropriate guidance.

Looking forward to hearing more about your case so that we can help you further.

**Comment by ziyuuuuuu** [comment]
Thank you for all people's help and replies. I actually started a new issue, and I got help from Glenn-Jocher there. So for the people who had same problem can check this:

[I have trained a custom yolov5 instance model, and i have 2 class present in the predict image, what should i do i can just present just 1 class in the predict image](https://github.com/ultralytics/yolov5/issues/12304#top)
#12304](url). 

I am quite new for the coding world. Thank you for y'all.

**Comment by glenn-jocher** [comment]
@ziyuuuuuu thank you for your participation and for reaching out for help. It's great to hear that you've started a new issue and received assistance from Glenn-Jocher. This demonstrates the collaborative nature of the YOLOv5 community.

For those encountering a similar problem, I recommend checking out the following link for further guidance on training a custom YOLOv5 instance segmentation model with multiple classes:
https://github.com/ultralytics/yolov5/issues/12304#top

Once again, thank you to everyone for their support and assistance.
